address.go

```diff
@@ -81,6 +118,37 @@ type Address interface {
 // When the address does not encode the network, such as in the case of a raw
 // public key, the address will be associated with the passed defaultNet.
 func DecodeAddress(addr string, defaultNet *chaincfg.Params) (Address, error) {
+
+	// Bech32 encoded segwit addresses start with an human-readable part
+	// (hrp) followed by 1. For Bitcoin mainnet the hrp is "bc", and for
+	// testnet it is "tb". If the address string matches the hrp found in
+	// the defaultNet params, we try to decode it as a segwit address.
+	hrp := defaultNet.Bech32HRPSegwit
+	prefix := hrp + "1"
+	if strings.HasPrefix(strings.ToLower(addr), prefix) {
```

Comment:
This section is incorrect.  It is an error to assume the caller is passing the current network.  The default network (hence the name `defaultNet`) only applies to addresses that do not have a network encoded with them (the only one that is true for currently is raw pay-to-pubkey addresses).

A bech32 address, by definition, must have an hrp that identifies the network.  See the code below which calls `chaincfg.IsPubKeyHashAddrID(netID)` to see how this is handled for old style addresses.  The same approach must be followed here since the network has to be decoded from the string and it must be done according to registered networks via the chain parameters.